### PR TITLE
fix receiver registered at onAttach

### PR DESCRIPTION
**Pull request by dodgex**
see https://github.com/excilys/androidannotations/issues/1132

### CODE REVIEW

File: AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/holder/ReceiverRegistrationHolder.java [hunk 1]
```diff
@@ -58,17 +58,25 @@ private JFieldVar createIntentFilterField(IntentFilterData intentFilterData) {
...
 		for (String dataScheme : intentFilterData.getDataSchemeSet()) {
-			initBody.invoke(intentFilterField, "addDataScheme").arg(dataScheme);
+			intentFilterTarget.invoke(intentFilterField, "addDataScheme").arg(dataScheme);
 		}
 
 		return intentFilterField;
 	}
 
+	private JBlock getIntentFilterTarget(IntentFilterData intentFilterData) {
+		if (intentFilterData.registerAt.equals(RegisterAt.OnAttachOnDetach) && holder instanceof EFragmentHolder) {
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/holder/ReceiverRegistrationHolder.java]: I think using polymorphism and method overrides is a better design. Maybe you could create this method in `EComponentHolder` which just returns init body, and override it in `EFragmentHolder` to return the needed block?

**Comment by WonderCsabo** [comment]
Can you add a test?

**Comment by dodgex** [comment]
yeah.

File: AndroidAnnotations/functional-test-1-5/src/test/java/org/androidannotations/test15/receiver/FragmentWithReceiverTest.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	FragmentManager fragmentManager;
+
+	@Before
+	public void setUp() {
+		fragment = new FragmentWithReceiver_();
+		startFragment(fragment);
+	}
+
+	@Test
+	public void defaultReceiverCalledTest() {
```
Comment (WonderCsabo) [inline AndroidAnnotations/functional-test-1-5/src/test/java/org/androidannotations/test15/receiver/FragmentWithReceiverTest.java]: Remove the `Test` suffix from the method names.

Comment (dodgex) [inline AndroidAnnotations/functional-test-1-5/src/test/java/org/androidannotations/test15/receiver/FragmentWithReceiverTest.java]: whoops. done

File: AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/holder/ReceiverRegistrationHolder.java (hunk 1 above)
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/holder/ReceiverRegistrationHolder.java]: @dodgex did you miss this?

Comment (dodgex) [inline AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/holder/ReceiverRegistrationHolder.java]: yeah. sorry. i'll update the PR

File: AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/holder/EFragmentHolder.java [hunk 3]
```diff
@@ -547,6 +549,15 @@ public JBlock getOnDetachBeforeSuperBlock() {
 	}
 
 	@Override
+	public JBlock getIntentFilterTarget(IntentFilterData intentFilterData) {
+		if (RegisterAt.OnAttachOnDetach.equals(intentFilterData.getRegisterAt())) {
+			return getOnAttachAfterSuperBlock();
+		} else {
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/holder/EFragmentHolder.java]: Remove the else block, return is enough. Also it maybe more flexible to return super.getIntentFilterTarget here.

File: AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/holder/EComponentHolder.java [hunk 4]
```diff
@@ -76,6 +77,10 @@ public JBlock getInitBody() {
 		return getInit().body();
 	}
 
+	public JBlock getIntentFilterTarget(IntentFilterData intentFilterData) {
```
Comment (yDelouis) [inline AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/holder/EComponentHolder.java]: `getIntentFilterTarget` should be a method of `HasReceiverRegistration`.
`EComponentHolder` does not implement this interface so the method should not be there.
Write this method in both `EActivityHolder`, `EServiceHolder` and `EFragmentHolder`. Even if the implementation is the same for `EActivityHolder` and `EServiceHolder`.

Comment (dodgex) [inline AndroidAnnotations/androidannotations/src/main/java/org/androidannotations/holder/EComponentHolder.java]: i decided to go with `getIntentFilterInitializationBlock` as it only initializes the intentFilter.

**Comment by WonderCsabo** [comment]
FInally merged, thanks!

**Pull request merged**